gui/wxpython/lmgr/frame.py

```diff
@@ -98,14 +97,24 @@ def __init__(
         self.displayIndex = 0          # index value for map displays and layer trees
         self.currentPage = None       # currently selected page for layer tree notebook
         self.currentPageNum = None       # currently selected page number for layer tree notebook
-        self.workspaceFile = workspace    # workspace file
-        self.workspaceChanged = False     # track changes in workspace
-        # if we are currently loading workspace to ignore some events
-        self.loadingWorkspace = False
+
         self.cwdPath = None               # current working directory
 
         wx.Frame.__init__(self, parent=parent, id=id, size=size,
                           style=style, **kwargs)
+
+        self._giface = LayerManagerGrassInterface(self)
+
+        # workspace manager and workspace signals
+        self.workspace_manager = WorkspaceManager(parent=self,
+                                                  workspaceFile=workspace,
+                                                  giface=self._giface)
+        self.workspace_manager.setDisplayIndex.connect(
+            lambda index: self.setDisplayIndex(index))
+        self.workspace_manager.setCurrentPage.connect(
+            lambda page: self.setCurrentPage(page))
+        self.workspace_manager.setTitle.connect(self._setTitle)
```

Comment:
The signals here and the new methods in giface add only a little to the separation of the classes. This, in connection with the `WriteWorkspaceFile` class already knowing all the details of lmgr and whole GUI, suggests that this new class should just do the same, i.e., accessing all the components, while reading the workspace. So, what I'm saying is, let's not change `WriteWorkspaceFile` for now, instead follow the design which is already there and that is individual widgets don't know about workspace, workspace know about every widget. I think this will allow you to finish this without touching every workspace-related class in GUI.

Reply:
I agree with that. Not I have the error about Loading. It still the as you remember I talked about it during our videocall:
![chyba](https://user-images.githubusercontent.com/49241681/111070972-56da1c80-84d4-11eb-98e8-5c79bf77c401.PNG)

I tried to change order of commands in GMFrame init a bit but it did not help... not sure where is the problem. I tried master and here it works fine (yet it is de facto the same, just place in a different file).

Reply:
I have already found the error.